proximal/examples/test_deconv_sv_psf.py

```diff
@@ -0,0 +1,153 @@
+
+        im_sim[:] = 0
+        for i in range(n_psf):
+            im_sim[:] += fftconvolve(im_src[i],
+                                     np.rot90(psf_mode[i], k=2),
+                                     mode='same')
+
+    # Estimate the largest eigenvalue of convT[ conv(x)] .
+    scale_factor = np.sum(psf_mode, axis=(1, 2)).max() * n_psf
+    blur_op = LinOpFactory(im_shape, (n_psf, *im_shape),
+                           forward,
+                           adjoint,
+                           norm_bound=scale_factor)
+
+    return blur_op
+
+
+def weighted_planewise_conv(G, W, im_shape, n_psf: int):
+    ''' Weighted planewise convolution.
+
+        For each PSF with small supports, apply corresponding weights to the
+        source image, then simulate the image blur with 2D shift-invariant
+        convolution.
+    '''
+    assert im_shape[0] * im_shape[1] == G.shape[0]
+    psf_mode = G[:, :n_psf].T.reshape((n_psf, *im_shape))
+    weight = W[:n_psf].reshape((n_psf, *im.shape))
+
+    def forward(im_src, im_sim):
+        assert np.all(im_src.shape == im_sim.shape[1:])
+        assert im_sim.shape[0] == n_psf
+
+        for i in range(n_psf):
+            im_sim[i, :] = fftconvolve(psf_mode[i],
+                                       im_src * weight[i],
+                                       mode='same')
+
+    def adjoint(im_src, im_sim):
+        assert im_src.shape[0] == n_psf
+        assert np.all(im_src.shape[1:] == im_sim.shape)
+
+        im_sim[:] = 0
+        for i in range(n_psf):
+            im_sim[:] += fftconvolve(
+                im_src[i], np.rot90(psf_mode[i], k=2), mode='same') * weight[i]
+
+    # Estimate the largest eigenvalue of convT[ conv(x)] .
+    scale_factor = np.sum(psf_mode, axis=(1, 2)).max() * n_psf * weight.max()
+    blur_op = LinOpFactory(im_shape, (n_psf, *im_shape),
+                           forward,
+                           adjoint,
+                           norm_bound=scale_factor)
+
+    return blur_op
+
+
+def NagyAndOLeary1998(x, im, U, W, mu=0.01, n_psf=15):
+    ''' Convolve first, then apply weights.
+
+        According to Eq. 23 of the review article (link below), there exists a
+        non-equivalent image interpolation formulation introduced by Nagy and
+        O'Leary.
+         
+        Reference: Denis, Thiebaut, Soulez et al. 2015,
+        https://doi.org/10.1007/s11263-015-0817-x
+    '''
+    blur_op = planewise_conv(U, im.shape, n_psf)
+    im_blur = blur_op(x)
+
+    weight = W[:n_psf].reshape((n_psf, *im.shape))
+    im_stack = np.empty(weight.shape, dtype=np.float32, order='F')
+    im_stack[:] = im[np.newaxis, ...]
+
+    # The weights are manually "absorbed" into the L2-norm, which has a fast
+    # elemwise proximal operator.
+    #
+    # TODO: Crop the im_stack and weight to double of image grid size; elminiate
+    # redundant 2D convolution with zero weights.
+    data_term = px.weighted_sum_squares(im_blur - im_stack, weight)
```

Comment:
ProxImaL "compiler" can work harder optimizing your problem if you let it absorb the weights into the proximal operator. Even better, if the weights has a small extent/support, we can crop and center each image "tiles" to further reduce the compute time.